statistics/handle/cache/internal/lfu/lfu_cache.go

```diff
@@ -0,0 +1,132 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lfu
+
+import (
+	"sync/atomic"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/pingcap/tidb/statistics"
+	"github.com/pingcap/tidb/statistics/handle/cache/internal"
+	"github.com/pingcap/tidb/util/intest"
+)
+
+const keySetCnt = 128
+
+func shard(key int64) int {
+	return int(key) % keySetCnt
+}
+
+// LFU is a LFU based on the ristretto.Cache
+type LFU struct {
+	cache        *ristretto.Cache
+	resultKeySet [keySetCnt]keySet
+	cost         atomic.Int64
+}
+
+// NewLFU creates a new LFU cache.
+func NewLFU(cost int64) (*LFU, error) {
+	result := &LFU{}
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters:        1024 * 1024 * 1024,
+		MaxCost:            cost,
+		BufferItems:        1,
+		OnEvict:            result.onEvict,
+		OnExit:             result.onExit,
+		IgnoreInternalCost: intest.InTest,
+	})
+	result.cache = cache
+	for i := 0; i < keySetCnt; i++ {
+		result.resultKeySet[i] = keySet{
+			set: make(map[int64]struct{}),
+		}
+	}
+	return result, err
+}
+
+// Get implements statsCacheInner
+func (s *LFU) Get(tid int64, _ bool) (*statistics.Table, bool) {
+	result, ok := s.cache.Get(tid)
+	if !ok {
+		return nil, ok
+	}
+	return result.(*statistics.Table), ok
+}
+
+// Put implements statsCacheInner
+func (s *LFU) Put(tblID int64, tbl *statistics.Table, _ bool) {
+	ok := s.cache.Set(tblID, tbl, tbl.MemoryUsage().TotalTrackingMemUsage())
+	if ok {
+		idx := shard(tblID)
+		s.resultKeySet[idx].Add(tblID)
+		s.cost.Add(tbl.MemoryUsage().TotalTrackingMemUsage())
+	}
+}
+
+// Del implements statsCacheInner
+func (s *LFU) Del(tblID int64) {
+	s.cache.Del(tblID)
+	idx := shard(tblID)
+	s.resultKeySet[idx].Remove(tblID)
+}
+
+// Cost implements statsCacheInner
+func (s *LFU) Cost() int64 {
+	return s.cost.Load()
+}
+
+// Values implements statsCacheInner
+func (s *LFU) Values() []*statistics.Table {
+	result := make([]*statistics.Table, 0, len(s.resultKeySet))
+	for key := range s.resultKeySet {
+		if value, ok := s.cache.Get(key); ok {
+			result = append(result, value.(*statistics.Table))
+		}
+	}
+	return nil
+}
+
+func (s *LFU) onEvict(item *ristretto.Item) {
+	idx := shard(int64(item.Key))
+	s.resultKeySet[idx].Remove(int64(item.Key))
+}
+
+func (s *LFU) onExit(val interface{}) {
+	s.cost.Add(-1 * val.(*statistics.Table).MemoryUsage().TotalTrackingMemUsage())
+}
+
+// Len implements statsCacheInner
+func (s *LFU) Len() int {
+	var result int
+	for i := 0; i < keySetCnt; i++ {
+		result += s.resultKeySet[i].Len()
+	}
+	return result
+}
+
+// Front implements statsCacheInner
+func (*LFU) Front() int64 {
+	return 0
+}
+
+// Copy implements statsCacheInner
+func (s *LFU) Copy() internal.StatsCacheInner {
+	return s
```

Comment:
Why not implement this method?

Reply:
LFU can not be copied. it can not copy its cm sketch. Copy interface will be deleted later。

Reply:
Can we implement a simple version of `copy` above the `LFU`? e.g. read all keys and values from the old cache and add them into a new cache.

Reply:
It can be implemented. But this copy is to COW. so if we do not use the COW, it is useless.